File: src/advent/2019/day1.clj
(ns advent.2019.day1
  "Advent of Code 2019, day 1: The Tyranny of the Rocket Equation"
  (:require [advent.helpers :as h]))

(def puzzle-input (h/slurp-resource "2019/day1.txt" h/slurp-int-lines))

(defn mass-of-module [module]
  (-> (/ module 3)
      (Math/floor)
      (- 2)
      (int)))

(defn combined-mass-of-module [module]
  (loop [tot 0 module module]
    (let [m (mass-of-module module)]
      (if (pos? m)
        (recur (+ tot m) m)
        tot))))

(defn puzzle1 [input]
  (apply + (map mass-of-module input)))

(defn puzzle2 [input]
  (apply + (map combined-mass-of-module input)))

File: src/rust/year2015/day1.rs
// Advent of Code 2015, day 1: Not Quite Lisp
// https://adventofcode.com/2015/day/1

fn floor_direction(i: u8) -> i32 {
    if i == 40 {
        1
    } else {
        -1
    }
}

fn puzzle1(input: &str) -> i32 {
    let mut floor = 0;
    for i in input.bytes() {
        floor += floor_direction(i);
    }
    floor
}

fn puzzle2(input: &str) -> i32 {
    let mut floor = 0;
    let mut idx = 0;
    for i in input.bytes() {
        idx += 1;
        floor += floor_direction(i);
        if floor < 0 {
            break;
        }
    }
    idx
}

pub fn run(input: &str) {
    println!("{:?}", puzzle1(input));
    println!("{:?}", puzzle2(input));
}
